streaming/in.go

```diff
@@ -1,28 +1,68 @@
-// Package streaming contains writer and reader implementing Crypt4GH encryption and decryption correspondingly.
+// Package streaming contains Writer and Reader implementing
+// Crypt4GH encryption and decryption correspondingly.
 package streaming
 
 import (
 	"bytes"
-	"container/list"
 	"errors"
 	"io"
+	"sync"
 
 	"github.com/neicnordic/crypt4gh/model/body"
 	"github.com/neicnordic/crypt4gh/model/headers"
 	"golang.org/x/crypto/chacha20poly1305"
 )
 
+// crypt4GHInternalReader is the internal structure for managing
+// the stream reader
 type crypt4GHInternalReader struct {
+	// reader is the Reader providing the encrypted stream data
+	// is consumed from.
 	reader io.Reader
 
-	header                                []byte
+	// header is a binary copy of the C4GH file header.
+	header []byte
+
+	// dataEncryptionParametersHeaderPackets may be one or more
+	// DataEncryptionParametersHeaderPacket:s. These provide e.g. symmetric
+	// keys for decrypting segments.
 	dataEncryptionParametersHeaderPackets []headers.DataEncryptionParametersHeaderPacket
-	dataEditList                          *headers.DataEditListHeaderPacket
-	encryptedSegmentSize                  int
-	lastDecryptedSegment                  int
-	buffer                                bytes.Buffer
+
+	// dataEditList possibly contains a pointer to a list of edits to apply
+	// (skip) when consuming the stream.
+	dataEditList *headers.DataEditListHeaderPacket
+
+	// encryptedSegmentSize is the size of a segment in the encrypted stream,
+	// i.e. 65536 (data)+any extras added such as MAC or nonce.
+	encryptedSegmentSize int
+
+	// lastDecryptedSegment is the number of the segment that was last decrypted
+	// (is available in buffer, if any).
+	lastDecryptedSegment int64
+
+	// buffer is where decrypted data is stored temporarily for consumption. It
+	// contains at most one segments worth of data.
+	buffer bytes.Buffer
+
+	// bufferUse is the size of the last segment put into buffer at the time of
+	// writing. bufferUse-buffer.Len() give the number of bytes consumed from
+	// the buffer already. (Go 1.21 introduces buffer.Available() which allows
```

Comment:
i assume we are not using `buffer.Available` or not sure i follow the idea here, is it to change it later ?

Reply:
Yes, it seems to early to bump to require 1.21 now (e.g. Ubuntu offers 1.20), so it's a note for the future. Grateful for better formulations to make that clearer if possible.